### PR TITLE
Bump cobra and add completions for fish

### PR DESCRIPTION
**Pull request by soraxas**
This PR bumps the version of `cobra` to utilise it's new completion feature for fish (introduced in v1.0.0).

What does this PR change? What problem does it solve?
-----------------------------------------------------

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: doc/fish-completion.fish [hunk 1]
```diff
@@ -0,0 +1,164 @@
+# fish completion for restic                               -*- shell-script -*-
```
Comment (MichaelEischer) [inline doc/fish-completion.fish:1]: That completion file should be generated during the release process. Please extend `helpers/prepare-release/main.go` accordingly and remove this file.

File: cmd/restic/cmd_generate.go [hunk 2]
```diff
@@ -38,6 +39,7 @@ func init() {
 	fs.StringVar(&genOpts.ManDir, "man", "", "write man pages to `directory`")
 	fs.StringVar(&genOpts.BashCompletionFile, "bash-completion", "", "write bash completion `file`")
 	fs.StringVar(&genOpts.ZSHCompletionFile, "zsh-completion", "", "write zsh completion `file`")
+	fs.StringVar(&genOpts.FishCompletionFile, "fish-completion", "", "write fish completion `file`")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_generate.go:42]: Please move that option above the zsh completion file

File: cmd/restic/cmd_generate.go [hunk 3]
```diff
@@ -10,10 +10,10 @@ import (
 
 var cmdGenerate = &cobra.Command{
 	Use:   "generate [flags]",
-	Short: "Generate manual pages and auto-completion files (bash, zsh)",
+	Short: "Generate manual pages and auto-completion files (bash, zsh, fish)",
 	Long: `
 The "generate" command writes automatically generated files (like the man pages
-and the auto-completion files for bash and zsh).
+and the auto-completion files for bash and zsh, fish).
```
Comment (MichaelEischer) [inline cmd/restic/cmd_generate.go:16]: This should be `bash, fish and zsh`.

File: cmd/restic/cmd_generate.go [hunk 4]
```diff
@@ -10,10 +10,10 @@ import (
 
 var cmdGenerate = &cobra.Command{
 	Use:   "generate [flags]",
-	Short: "Generate manual pages and auto-completion files (bash, zsh)",
+	Short: "Generate manual pages and auto-completion files (bash, zsh, fish)",
```
Comment (MichaelEischer) [inline cmd/restic/cmd_generate.go:13]: Please sort the shells alphabetically

File: cmd/restic/cmd_generate.go [hunk 5]
```diff
@@ -68,6 +70,11 @@ func writeZSHCompletion(file string) error {
 	return cmdRoot.GenZshCompletionFile(file)
 }
 
+func writeFishCompletion(file string) error {
```
Comment (MichaelEischer) [inline cmd/restic/cmd_generate.go:73]: Please move that function above the one for zsh.

File: cmd/restic/cmd_generate.go [hunk 6]
```diff
@@ -90,6 +97,13 @@ func runGenerate(cmd *cobra.Command, args []string) error {
 		}
 	}
 
+	if genOpts.FishCompletionFile != "" {
```
Comment (MichaelEischer) [inline cmd/restic/cmd_generate.go:100]: Same here

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
This PR needs further work:
- The `fish-completion.fish` should be autogenerated on release.
- In addition the way the file content currently looks, I don't understand why / whether it works. Is there some fish magic involved?
- There needs to be documentation on how to use / install the completions.
- Please add a changelog entry.
- Please fix the linter error.

I'm not too thrilled to update cobra as that leads to a way uglier manpage formatting than before. But cobra 1.1.3 seems to have at least fixed the manpage headers.

Does anyone know how to correctly update `contrib/restic.spec`?

**Comment by soraxas** [comment]
Hi @MichaelEischer thank you for your review and detail instruction. I've addressed all of the mentioned issues, including amending `contrib/restic.spec` for installing the completion file.

Regarding your second question, it's cobra magic where it has a hidden internal subcommand `<CMD>  __complete` (e.g. `restic __complete`) where it's used by the completion scripts to auto-complete.

With the current lint failed test, I can't seems to pin point what's causing it. I've ran `go mod tidy` on my local machine and committed that change but it's still complaining.

**Comment by MichaelEischer** [comment]
Sorry, for the late reply. Could you squash the commits and then rebase them to the current master branch? Running `go mod tidy` with go 1.16 should fix the lint errors.

**Comment by soraxas** [comment]
Yep squashed and rebased

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**